cli/tests/pcluster/utils.py

```diff
@@ -65,6 +65,38 @@ def get_asset_content_with_resource_name(assets, resource_name):
     return None
 
 
+def get_resource_from_assets(assets, resource_name):
+    """Return the Resource dictionary containing a specific resource."""
+    asset = get_asset_content_with_resource_name(assets, resource_name)
+    if asset:
+        return asset["Resources"][resource_name]
+
+
+# It expects as input ['UserData']['Fn::Base64']['Fn::Sub'] that is a list of two elements
+# 1: contains a dictionary with the key-values to be replaced by Fn::Sub
+# 0: contains the file (as str) where those values will be replaced
+def render_user_data(user_data):
+    """Render the UserData as CF will do."""
+    content = user_data[0]
+    values = user_data[1]
+
+    for value in values:
+        content = content.replace("${%s}" % value, str(values[value]))
+
+    return content
+
+
+def validate_dna_json_fields(user_data, fields):
+    """Validate that all dna.json fields, expressed as key-value pairs, are present in the user_data."""
+    lines = user_data.splitlines()
+    for k, v in fields.items():
+        for line in lines:
+            if k in line:
+                break
+
+        assert v in line, f"Validating Key '{k}' - Value: '{v}' not found in dna.json line:  '{line}'"
+
+
```

Comment:
Maybe I'm missing some context here, but why can't we import the rendered dna.json and the expected one as dictionaries and compare them via [`assertpy.is_subset_of`](https://assertpy.github.io/docs.html#assertpy.collection.CollectionMixin.is_subset_of)?

Reply:
For the HeadNode we are rendering the dna.json as a single file and so we can read it again as a dictionary.
For the LoginNodes (and ComputeNodes) the dna.json is a hidden script in the user_data that will create a file in the remote host while running.

Extracting this section from the user_data, creating a temporary file and re-reading it as a dictionary is something we can do ..but as a future improvement.